### PR TITLE
[PM-6909] Centralize database migration logic

### PR DESCRIPTION
**Pull request by withinfocus**
## Objective

Centralizes database migration logic.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: util/Migrator/DbMigrator.cs [hunk 1]
```diff
@@ -12,39 +12,34 @@ public class DbMigrator
 {
     private readonly string _connectionString;
     private readonly ILogger<DbMigrator> _logger;
-    private readonly string _masterConnectionString;
 
-    public DbMigrator(string connectionString, ILogger<DbMigrator> logger)
```
Comment (withinfocus) [inline util/Migrator/DbMigrator.cs:17]: ℹ️ The logger is now created internally. Three callsites with no customization and two without DI.

File: util/Migrator/DbMigrator.cs [hunk 2]
```diff
@@ -12,39 +12,34 @@ public class DbMigrator
...
-        CancellationToken cancellationToken = default(CancellationToken))
+        CancellationToken cancellationToken = default)
     {
         var attempt = 1;
-
         while (attempt < 10)
         {
             try
             {
+                PrepareDatabase(cancellationToken);
```
Comment (withinfocus) [inline util/Migrator/DbMigrator.cs:32]: ℹ️ Separation of "preparation" tasks from the actual migration. The tasks in here really don't have to run any time but once, but not going after that here.

File: util/Migrator/DbMigrator.cs [hunk 3]
```diff
@@ -56,17 +51,14 @@ public DbMigrator(string connectionString, ILogger<DbMigrator> logger)
         return false;
     }
 
-    public bool MigrateDatabase(bool enableLogging = true,
-        bool repeatable = false,
-        string folderName = MigratorConstants.DefaultMigrationsFolderName,
-        CancellationToken cancellationToken = default(CancellationToken))
+    private void PrepareDatabase(CancellationToken cancellationToken = default)
     {
-        if (_logger != null)
```
Comment (withinfocus) [inline util/Migrator/DbMigrator.cs:64]: ℹ️ This is now lower down.

❓ There was a confusing mix of logger usages in this class so I just went with "use it or not". The only project not logging is `Setup` -- why? Do we want to suppress setup log messages?

File: util/Migrator/DbMigrator.cs [hunk 4]
```diff
@@ -119,20 +126,13 @@ public DbMigrator(string connectionString, ILogger<DbMigrator> logger)
...
         if (enableLogging)
         {
-            if (_logger != null)
-            {
-                builder.LogTo(new DbUpLogger(_logger));
-            }
-            else
-            {
-                builder.LogToConsole();
-            }
```
Comment (withinfocus) [inline util/Migrator/DbMigrator.cs:129]: ℹ️ This seems to be what our logger will do ...

File: util/Migrator/SqlServerDbMigrator.cs [hunk 5]
```diff
@@ -1,109 +1,21 @@
...
-using Microsoft.Extensions.Logging;
 
 namespace Bit.Migrator;
 
 public class SqlServerDbMigrator : IDbMigrator
 {
-    private readonly string _connectionString;
-    private readonly ILogger<SqlServerDbMigrator> _logger;
-    private readonly string _masterConnectionString;
+    private readonly DbMigrator _migrator;
```
Comment (withinfocus) [inline util/Migrator/SqlServerDbMigrator.cs:8]: ❓ Everything in this class was duplicated by the core migrator, so now it just uses that like `Setup` and the MSSQL utility. Am I missing anything? Why was this copied?

File: util/MsSqlMigratorUtility/Program.cs [hunk 6]
```diff
@@ -15,60 +12,26 @@ private static int Main(string[] args)
     public void Execute(
         [Operand(Description = "Database connection string")]
         string databaseConnectionString,
-        [Option('v', "verbose", Description = "Enable verbose output of migrator logs")]
```
Comment (withinfocus) [inline util/MsSqlMigratorUtility/Program.cs:18]: ❓ Nothing emitted debug messages so this seemed unnecessary. Am I missing something here? Do we use this anywhere, like with automation?

File: util/Setup/Program.cs [hunk 7]
```diff
@@ -185,17 +186,19 @@ private static void PrintEnvironment()
...
-    private static void MigrateDatabase(int attempt = 1)
+    private static void PrepareAndMigrateDatabase()
     {
         var vaultConnectionString = Helpers.GetValueFromEnvFile("global",
             "globalSettings__sqlServer__connectionString");
-        var migrator = new DbMigrator(vaultConnectionString, null);
+        var migrator = new DbMigrator(vaultConnectionString);
 
-        var log = false;
+        var enableLogging = false;
```
Comment (withinfocus) [inline util/Setup/Program.cs:195]: ℹ️ If we determine this is unnecessary I can kill off the whole parameter down the chain.

**Comment by withinfocus** [comment]
Added a few reviewers directly with historical knowledge and for any gotchas.

File: util/MsSqlMigratorUtility/Program.cs (hunk 6 above)
Comment (michalchecinski) [reply util/MsSqlMigratorUtility/Program.cs:18]: It's used in the automations. If the `verbose` flag is set during execution, the Migrator logs the SQL scripts applied during the run. If we want to remove the `verbose` option, I would like to set the `enableLogging` param in the `MigrateMsSqlDatabaseWithRetries` method to always be `true`, like you did in lines 28 and 32. This is helpful to see which scripts are run in the automation (to confirm if all needed have been applied).

Comment (withinfocus) [reply util/MsSqlMigratorUtility/Program.cs:18]: OK, so even though you invoked this as "verbose", what it was actually doing is simply logging. It configured a debug-level log emission, but nothing logs at that level. I think this is fine then and at some point the automations could omit that parameter once this merges.

You can see in the Action run logs for this PR that you still get those migration messages. Please review the database testing run if you don't mind to ensure this works as you (DevOps) expect.

Comment (michalchecinski) [reply util/MsSqlMigratorUtility/Program.cs:18]: I just checked it, and it works as expected 😃

Comment (michalchecinski) [reply util/MsSqlMigratorUtility/Program.cs:18]: @withinfocus We need to ensure that this change is backwards compatible. If we use the `-v` flag after removing the verbose option, will the `MsSqlMigratoeUtility` run as expected, or will it throw an error stating that it does not know the `-v` option?

Comment (withinfocus) [reply util/MsSqlMigratorUtility/Program.cs:18]: Sigh, I hacked around on this and can't get `Unrecognized option '-v'` to leave. I pushed https://github.com/bitwarden/server/pull/3910/commits/32b0801a6ebfbb8b54dc901126822af3a277d966 just now for this -- is that all that's needed?

**Review by rkac-bw — DISMISSED** [review]
LGTM

File: util/MsSqlMigratorUtility/Program.cs (hunk 6 above)
Comment (michalchecinski) [reply util/MsSqlMigratorUtility/Program.cs:18]: The `${@}` should stay, as it passes other args, e.g. the `--repeatable` or `--folder`.

File: util/MsSqlMigratorUtility/Dockerfile [hunk 8]
```diff
@@ -5,4 +5,4 @@ LABEL com.bitwarden.product="bitwarden"
 WORKDIR /app
 COPY obj/build-output/publish .
 
-ENTRYPOINT ["sh", "-c", "dotnet /app/MsSqlMigratorUtility.dll \"${MSSQL_CONN_STRING}\" -v ${@}", "--" ]
+ENTRYPOINT ["sh", "-c", "dotnet /app/MsSqlMigratorUtility.dll \"${MSSQL_CONN_STRING}\"", "--" ]
```
Comment (michalchecinski) [inline util/MsSqlMigratorUtility/Dockerfile:8]:
```suggestion
ENTRYPOINT ["sh", "-c", "dotnet /app/MsSqlMigratorUtility.dll \"${MSSQL_CONN_STRING}\" ${@}", "--" ]
```

**Review by michalchecinski — CHANGES_REQUESTED** [review]

**Review by rkac-bw — DISMISSED** [review]

**Review by michalchecinski — DISMISSED** [review]

File: util/Migrator/DbMigrator.cs (hunk 1 above)
Comment (justindbaur) [reply util/Migrator/DbMigrator.cs:17]: `SqlServerDbMigrator` got theirs from DI which could be customized, how about leave a constructor that takes it but offer one without which will just create the default one?

Comment (withinfocus) [reply util/Migrator/DbMigrator.cs:17]: Lemme know if https://github.com/bitwarden/server/pull/3910/commits/7d1eee7ae3a3f8ac5c45ce73263e86eeb60a5f37 works for you.

Comment (justindbaur) [reply util/Migrator/DbMigrator.cs:17]: Perfect

**Review by rkac-bw — APPROVED** [review]
LGTM

**Review by michalchecinski — APPROVED** [review]

**Pull request merged**